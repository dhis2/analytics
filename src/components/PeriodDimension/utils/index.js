export const DAILY = 'DAILY'
export const WEEKLY = 'WEEKLY'
export const WEEKLYWED = 'WEEKLYWED'
export const WEEKLYTHU = 'WEEKLYTHU'
export const WEEKLYSAT = 'WEEKLYSAT'
export const WEEKLYSUN = 'WEEKLYSUN'
export const WEEKS_THIS_YEAR = 'WEEKS_THIS_YEAR'
export const BIWEEKLY = 'BIWEEKLY'
export const MONTHLY = 'MONTHLY'
export const BIMONTHLY = 'BIMONTHLY'
export const QUARTERLY = 'QUARTERLY'
export const SIXMONTHLY = 'SIXMONTHLY'
export const SIXMONTHLYAPR = 'SIXMONTHLYAPR'
export const YEARLY = 'YEARLY'
export const FINANCIAL = 'FINANCIAL'
export const FYNOV = 'FYNOV'
export const FYOCT = 'FYOCT'
export const FYJUL = 'FYJUL'
export const FYAPR = 'FYAPR'

export const filterPeriodTypesById = (
    allPeriodTypes = [],
    excludedPeriodTypes = []
) => allPeriodTypes.filter(period => !excludedPeriodTypes.includes(period.id))
