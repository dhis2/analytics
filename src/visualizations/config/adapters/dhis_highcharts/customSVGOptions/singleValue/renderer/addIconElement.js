const parser = new DOMParser()

export function addIconElement(svgString, color) {
    const svgIconDocument = parser.parseFromString(svgString, 'image/svg+xml')
    const iconElHeight = svgIconDocument.documentElement.getAttribute('height')
    const iconElWidth = svgIconDocument.documentElement.getAttribute('width')
    const iconGroup = this.renderer
        .g('icon')
        .attr('data-test', 'visualization-icon')
        .css({
            color,
        })
    /* Force the group element to have the same dimensions as the original
     * SVG image by adding this rect. This ensures the icon has the intended
     * whitespace around it and makes scaling and translating easier. */
    this.renderer.rect(0, 0, iconElWidth, iconElHeight).add(iconGroup)

    Array.from(svgIconDocument.documentElement.children).forEach((node) =>
        iconGroup.element.appendChild(node)
    )

    iconGroup.add()

    return iconGroup
}
