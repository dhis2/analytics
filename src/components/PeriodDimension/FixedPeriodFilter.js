import React from 'react'
import PropTypes from 'prop-types'
import i18n from '@dhis2/d2-i18n'
import {
    SingleSelectField,
    SingleSelectOption,
    InputField,
} from '@dhis2/ui-core'

import { getFixedPeriodsOptions } from './utils/fixedPeriods'
import styles from './styles/PeriodFilter.style'

const FixedPeriodFilter = ({
    currentPeriodType,
    currentYear,
    onSelectPeriodType,
    onSelectYear,
}) => (
    <>
        <div className="leftSection">
            <SingleSelectField
                label={i18n.t('Period type')}
                onChange={({ selected }) => onSelectPeriodType(selected)}
                dense
                selected={currentPeriodType}
                className="filterElement"
            >
                {getFixedPeriodsOptions().map(option => (
                    <SingleSelectOption
                        key={option.id}
                        value={option.id}
                        label={option.getName()}
                    />
                ))}
            </SingleSelectField>
        </div>
        <div className="rightSection">
            <InputField
                label={i18n.t('Year')}
                className="filterElement"
                type="number"
                placeholder={i18n.t('Select year')}
                value={currentYear}
                onChange={({ value }) => onSelectYear(value)}
                dense
            ></InputField>
        </div>
        <style jsx>{styles}</style>
    </>
)

FixedPeriodFilter.propTypes = {
    currentPeriodType: PropTypes.object.isRequired,
    currentYear: PropTypes.string.isRequired,
    onSelectPeriodType: PropTypes.func.isRequired,
    onSelectYear: PropTypes.func.isRequired,
}

export default FixedPeriodFilter
