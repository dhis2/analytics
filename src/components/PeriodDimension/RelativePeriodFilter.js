import React from 'react'
import PropTypes from 'prop-types'
import i18n from '@dhis2/d2-i18n'
import { SingleSelectOption, SingleSelectField } from '@dhis2/ui-core'

import { getRelativePeriodsOptions } from './utils/relativePeriods'
import styles from './styles/PeriodFilter.style'

const RelativePeriodFilter = ({ currentFilter, onSelectFilter }) => (
    <div className="leftSection">
        <SingleSelectField
            label={i18n.t('Period type')}
            onChange={({ selected }) => onSelectFilter(selected)}
            dense
            selected={currentFilter}
            className="filterElement"
        >
            {getRelativePeriodsOptions().map(option => (
                <SingleSelectOption
                    key={option.id}
                    value={option.id}
                    label={option.getName()}
                />
            ))}
        </SingleSelectField>
        <style jsx>{styles}</style>
    </div>
)

RelativePeriodFilter.propTypes = {
    currentFilter: PropTypes.object.isRequired,
    onSelectFilter: PropTypes.func.isRequired,
}

export default RelativePeriodFilter
