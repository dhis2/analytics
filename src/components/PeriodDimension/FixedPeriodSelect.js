import { SingleSelectField, SingleSelectOption } from '@dhis2/ui'
import PropTypes from 'prop-types'
import React, { Component } from 'react'
import i18n from '../../locales/index.js'
import FixedPeriodFilter from './FixedPeriodFilter.js'
import styles from './styles/FixedPeriodSelect.style.js'
import {
    parsePeriodCode,
    getFixedPeriodsOptionsById as getPeriodById,
    getYearOffsetFromNow,
} from './utils/fixedPeriods.js'

class FixedPeriodSelect extends Component {
    state = {
        periodType: '',
        year: '',
        options: null,
    }

    static getDerivedStateFromProps({ value, allowedPeriodTypes }) {
        const period = parsePeriodCode(value, allowedPeriodTypes)

        if (!period) {
            return null
        }

        return {
            periodType: period.id,
            year: period.year,
            options: period.options,
        }
    }

    onSelectPeriodType = (periodType) => {
        this.setState({
            periodType,
            options: this.getUpdatedOptions(periodType, this.state.year),
        })
        this.props.onChange()
    }

    onSelectYear = (year) => {
        this.setState({
            year,
            options: this.getUpdatedOptions(this.state.periodType, year),
        })
        this.props.onChange()
    }

    onSelectPeriod = ({ selected: periodId }, event) => {
        const period = this.state.options.find(({ id }) => id === periodId)
        this.props.onChange(period, event)
    }

    getUpdatedOptions(periodType, year) {
        const offset = getYearOffsetFromNow(year)
        return getPeriodById(periodType).getPeriods({ offset })
    }

    render() {
        return (
            <div
                data-test={this.props.dataTest}
                className={this.props.className}
            >
                <div className="row">
                    <FixedPeriodFilter
                        allowedPeriodTypes={this.props.allowedPeriodTypes}
                        currentPeriodType={this.state.periodType}
                        currentYear={this.state.year}
                        onSelectPeriodType={this.onSelectPeriodType}
                        onSelectYear={this.onSelectYear}
                    />
                </div>

                {this.state.periodType && this.state.year && (
                    <SingleSelectField
                        label={i18n.t('Period')}
                        onChange={this.onSelectPeriod}
                        dense
                        selected={this.props.value}
                    >
                        {this.state.options.map((option) => (
                            <SingleSelectOption
                                key={option.id}
                                value={option.id}
                                label={option.name}
                            />
                        ))}
                    </SingleSelectField>
                )}
                <style jsx>{styles}</style>
            </div>
        )
    }
}

FixedPeriodSelect.defaultProps = {
    dataTest: 'dhis2-analytics-fixedperiodselect',
    value: '',
}

FixedPeriodSelect.propTypes = {
    onChange: PropTypes.func.isRequired,
    allowedPeriodTypes: PropTypes.arrayOf(PropTypes.string),
    className: PropTypes.string,
    dataTest: PropTypes.string,
    value: PropTypes.string,
}

export default FixedPeriodSelect
