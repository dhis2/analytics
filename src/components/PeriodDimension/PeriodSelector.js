import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { Transfer, TabBar, Tab } from '@dhis2/ui-core'
import i18n from '@dhis2/d2-i18n'

import FixedPeriodFilter from './FixedPeriodFilter'
import RelativePeriodFilter from './RelativePeriodFilter'
import { MONTHS, getRelativePeriodsOptionsById } from './utils/relativePeriods'
import { MONTHLY, getFixedPeriodsOptionsById } from './utils/fixedPeriods'
import styles from '../styles/DimensionSelector.style'
import { TransferOption } from '../TransferOption'
import PeriodIcon from '../../assets/DimensionItemIcons/PeriodIcon'
import {
    TRANSFER_HEIGHT,
    TRANSFER_OPTIONS_WIDTH,
    TRANSFER_SELECTED_WIDTH,
} from '../../modules/dimensionSelectorHelper'

const defaultRelativePeriodType = getRelativePeriodsOptionsById(MONTHS)
const defaultFixedPeriodType = getFixedPeriodsOptionsById(MONTHLY)
const defaultFixedPeriodYear = new Date().getFullYear()
const fixedPeriodConfig = year => ({
    offset: year - defaultFixedPeriodYear,
    filterFuturePeriods: false,
    reversePeriods: false,
})

class PeriodSelector extends Component {
    state = {
        allPeriods: defaultRelativePeriodType.getPeriods(),
        selectedPeriods: [],
        isRelative: true,
        relativeFilter: {
            periodType: {
                label: defaultRelativePeriodType.getName(),
                value: defaultRelativePeriodType.id,
            },
        },
        fixedFilter: {
            periodType: {
                label: defaultFixedPeriodType.getName(),
                value: defaultFixedPeriodType.id,
            },
            year: defaultFixedPeriodYear.toString(),
        },
    }

    constructor(props) {
        super(props)

        this.state.selectedPeriods = this.props.initialSelectedPeriods.map(
            item => ({
                label: item.name,
                value: item.id,
                key: item.id,
            })
        )
    }

    onIsRelativeClick = isRelative => {
        if (this.state.isRelative !== isRelative) {
            this.setState({ isRelative })
            this.setState({
                allPeriods: isRelative
                    ? getRelativePeriodsOptionsById(
                          this.state.relativeFilter.periodType.value
                      ).getPeriods()
                    : getFixedPeriodsOptionsById(
                          this.state.fixedFilter.periodType.value
                      ).getPeriods(
                          fixedPeriodConfig(Number(this.state.fixedFilter.year))
                      ),
            })
        }
    }

    renderHeader = () => (
        <>
            <TabBar>
                <Tab
                    selected={this.state.isRelative}
                    onClick={() => this.onIsRelativeClick(true)}
                >
                    {i18n.t('Relative periods')}
                </Tab>
                <Tab
                    selected={!this.state.isRelative}
                    onClick={() => this.onIsRelativeClick(false)}
                >
                    {i18n.t('Fixed periods')}
                </Tab>
            </TabBar>
            <div className="filterContainer">
                {this.state.isRelative ? (
                    <RelativePeriodFilter
                        currentFilter={this.state.relativeFilter.periodType}
                        onSelectFilter={filter => {
                            this.setState({
                                relativeFilter: { periodType: filter },
                            })
                            this.setState({
                                allPeriods: getRelativePeriodsOptionsById(
                                    filter.value
                                ).getPeriods(),
                            })
                        }}
                    />
                ) : (
                    <FixedPeriodFilter
                        currentPeriodType={this.state.fixedFilter.periodType}
                        currentYear={this.state.fixedFilter.year}
                        onSelectPeriodType={periodType => {
                            this.onSelectFixedPeriods({
                                periodType,
                                year: this.state.fixedFilter.year,
                            })
                        }}
                        onSelectYear={year => {
                            this.onSelectFixedPeriods({
                                periodType: this.state.fixedFilter.periodType,
                                year,
                            })
                        }}
                    />
                )}
            </div>
            <style jsx>{styles}</style>
        </>
    )

    onSelectFixedPeriods = fixedFilter => {
        this.setState({
            fixedFilter,
            allPeriods: getFixedPeriodsOptionsById(
                fixedFilter.periodType.value
            ).getPeriods(fixedPeriodConfig(Number(fixedFilter.year))),
        })
    }

    renderEmptySelection = () => (
        <>
            <p className="emptySelection">{i18n.t('No periods selected')}</p>
            <style jsx>{styles}</style>
        </>
    )

    render = () => (
        <Transfer
            onChange={({ selected }) => {
                this.setState({ selectedPeriods: selected })
                this.props.onSelect(selected)
            }}
            selected={this.state.selectedPeriods}
            leftHeader={this.renderHeader()}
            enableOrderChange
            height={TRANSFER_HEIGHT}
            optionsWidth={TRANSFER_OPTIONS_WIDTH}
            selectedWidth={TRANSFER_SELECTED_WIDTH}
            selectedEmptyComponent={this.renderEmptySelection()}
            rightFooter={this.props.rightFooter}
            // TODO: Add rightHeader "Selected Periods" once the Transfer component supports this (https://github.com/dhis2/ui-core/issues/885)
        >
            {this.state.allPeriods.map(item => (
                <TransferOption
                    label={item.getName()}
                    value={item.id}
                    key={item.id}
                    icon={PeriodIcon}
                />
            ))}
        </Transfer>
    )
}

PeriodSelector.propTypes = {
    onSelect: PropTypes.func.isRequired,
    initialSelectedPeriods: PropTypes.arrayOf(PropTypes.object),
    rightFooter: PropTypes.node,
}

PeriodSelector.defaultProps = {
    initialSelectedPeriods: [],
}

export default PeriodSelector
