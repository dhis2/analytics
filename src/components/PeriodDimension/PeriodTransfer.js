import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { TabBar, Tab, Transfer } from '@dhis2/ui'
import i18n from '@dhis2/d2-i18n'

import FixedPeriodFilter from './FixedPeriodFilter'
import RelativePeriodFilter from './RelativePeriodFilter'
import { MONTHS, getRelativePeriodsOptionsById } from './utils/relativePeriods'
import { MONTHLY, getFixedPeriodsOptionsById } from './utils/fixedPeriods'
import styles from '../styles/DimensionSelector.style'
import { TransferOption } from '../TransferOption'
import PeriodIcon from '../../assets/DimensionItemIcons/PeriodIcon' //TODO: Reimplement the icon
import {
    TRANSFER_HEIGHT,
    TRANSFER_OPTIONS_WIDTH,
    TRANSFER_SELECTED_WIDTH,
} from '../../modules/dimensionSelectorHelper'

const defaultRelativePeriodType = getRelativePeriodsOptionsById(MONTHS)
const defaultFixedPeriodType = getFixedPeriodsOptionsById(MONTHLY)
const defaultFixedPeriodYear = new Date().getFullYear()
const fixedPeriodConfig = year => ({
    offset: year - defaultFixedPeriodYear,
    filterFuturePeriods: false,
    reversePeriods: false,
})

class PeriodTransfer extends Component {
    state = {
        allPeriods: defaultRelativePeriodType.getPeriods(),
        selectedPeriods: [],
        isRelative: true,
        relativeFilter: {
            periodType: defaultRelativePeriodType.id,
        },
        fixedFilter: {
            periodType: defaultFixedPeriodType.id,
            year: defaultFixedPeriodYear.toString(),
        },
    }

    constructor(props) {
        super(props)

        this.state.selectedPeriods = this.props.initialSelectedPeriods
    }

    onIsRelativeClick = isRelative => {
        if (this.state.isRelative !== isRelative) {
            this.setState({ isRelative })
            this.setState({
                allPeriods: isRelative
                    ? getRelativePeriodsOptionsById(
                          this.state.relativeFilter.periodType
                      ).getPeriods()
                    : getFixedPeriodsOptionsById(
                          this.state.fixedFilter.periodType
                      ).getPeriods(
                          fixedPeriodConfig(Number(this.state.fixedFilter.year))
                      ),
            })
        }
    }

    renderLeftHeader = () => (
        <>
            <TabBar>
                <Tab
                    selected={this.state.isRelative}
                    onClick={() => this.onIsRelativeClick(true)}
                >
                    {i18n.t('Relative periods')}
                </Tab>
                <Tab
                    selected={!this.state.isRelative}
                    onClick={() => this.onIsRelativeClick(false)}
                >
                    {i18n.t('Fixed periods')}
                </Tab>
            </TabBar>
            <div className="filterContainer">
                {this.state.isRelative ? (
                    <RelativePeriodFilter
                        currentFilter={this.state.relativeFilter.periodType}
                        onSelectFilter={filter => {
                            this.setState({
                                relativeFilter: { periodType: filter },
                            })
                            this.setState({
                                allPeriods: getRelativePeriodsOptionsById(
                                    filter
                                ).getPeriods(),
                            })
                        }}
                    />
                ) : (
                    <FixedPeriodFilter
                        currentPeriodType={this.state.fixedFilter.periodType}
                        currentYear={this.state.fixedFilter.year}
                        onSelectPeriodType={periodType => {
                            this.onSelectFixedPeriods({
                                periodType,
                                year: this.state.fixedFilter.year,
                            })
                        }}
                        onSelectYear={year => {
                            this.onSelectFixedPeriods({
                                periodType: this.state.fixedFilter.periodType,
                                year,
                            })
                        }}
                    />
                )}
            </div>
            <style jsx>{styles}</style>
        </>
    )

    renderRightHeader = () => (
        <>
            <p className="rightHeader">{i18n.t('Selected Periods')}</p>
            <style jsx>{styles}</style>
        </>
    )

    onSelectFixedPeriods = fixedFilter => {
        this.setState({
            fixedFilter,
            allPeriods: getFixedPeriodsOptionsById(
                fixedFilter.periodType
            ).getPeriods(fixedPeriodConfig(Number(fixedFilter.year))),
        })
    }

    renderEmptySelection = () => (
        <>
            <p className="emptyList">{i18n.t('No periods selected')}</p>
            <style jsx>{styles}</style>
        </>
    )

    render = () => (
        <Transfer
            onChange={({ selected }) => {
                const formattedItems = selected.map(id => ({
                    id,
                    name: [
                        ...this.state.allPeriods,
                        ...this.state.selectedPeriods,
                    ].find(item => item.id === id).name,
                }))
                this.setState({ selectedPeriods: formattedItems })
                this.props.onSelect(formattedItems)
            }}
            selected={this.state.selectedPeriods.map(period => period.id)}
            leftHeader={this.renderLeftHeader()}
            enableOrderChange
            height={TRANSFER_HEIGHT}
            optionsWidth={TRANSFER_OPTIONS_WIDTH}
            selectedWidth={TRANSFER_SELECTED_WIDTH}
            selectedEmptyComponent={this.renderEmptySelection()}
            rightHeader={this.renderRightHeader()}
            rightFooter={this.props.rightFooter}
            options={[
                ...this.state.allPeriods,
                ...this.state.selectedPeriods,
            ].map(({ id, name }) => ({
                label: name,
                value: id,
            }))}
            renderOption={props => (
                <TransferOption {...props} icon={PeriodIcon} />
            )}
        ></Transfer>
    )
}

PeriodTransfer.propTypes = {
    onSelect: PropTypes.func.isRequired,
    initialSelectedPeriods: PropTypes.arrayOf(
        PropTypes.shape({
            id: PropTypes.string,
            name: PropTypes.string,
        })
    ),
    rightFooter: PropTypes.node,
}

PeriodTransfer.defaultProps = {
    initialSelectedPeriods: [],
}

export default PeriodTransfer
