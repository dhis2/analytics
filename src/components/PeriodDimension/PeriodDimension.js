import React from 'react'
import PropTypes from 'prop-types'

import PeriodSelector from './PeriodSelector'
import { DIMENSION_ID_PERIOD } from '../../modules/predefinedDimensions'

export const PeriodDimension = ({ onSelect, selectedPeriods, rightFooter }) => {
    const selectItems = periods => {
        const formattedPeriods = periods.map(period => ({
            id: period.value,
            name: period.label,
        }))
        onSelect({
            dimensionId: DIMENSION_ID_PERIOD,
            items: formattedPeriods,
        })
    }

    return (
        <PeriodSelector
            onSelect={selectItems}
            initialSelectedPeriods={selectedPeriods}
            rightFooter={rightFooter}
        />
    )
}

PeriodDimension.propTypes = {
    onSelect: PropTypes.func.isRequired,
    rightFooter: PropTypes.node,
    selectedPeriods: PropTypes.array,
}

PeriodDimension.defaultProps = {
    selectedPeriods: [],
}

export default PeriodDimension
