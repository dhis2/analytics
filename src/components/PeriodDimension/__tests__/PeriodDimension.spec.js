import React from 'react'
import renderer from 'react-test-renderer'
import { PeriodDimension } from '../PeriodDimension'

jest.mock('../../PeriodSelector/PeriodSelector', () => ({
    __esModule: true,
    default: 'mockPeriodSelector',
    namedExport: jest.fn(),
}))

describe('The Period Dimension component', () => {
    const props = {
        selectedPeriods: [],
        onSelect: jest.fn(),
        onDeselect: jest.fn(),
        onReorder: jest.fn(),
        /*
        ui: {
            itemsByDimension: {
                [peId]: [],
            },
        },
        metadata: {},
        addMetadata: jest.fn(),
        addUiItems: jest.fn(),
        removeUiItems: jest.fn(),
        setUiItems: jest.fn(),
        context: { d2: {} },
        */
    }

    it('renders correctly', () => {
        const tree = renderer.create(<PeriodDimension {...props} />).toJSON()
        expect(tree).toMatchSnapshot()
    })
})
