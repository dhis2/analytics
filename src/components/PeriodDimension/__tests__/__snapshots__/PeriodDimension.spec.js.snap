// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Period Dimension component matches the snapshot 1`] = `
<PeriodSelector
  initialSelectedPeriods={Array []}
  onSelect={[Function]}
  rightFooter={<React.Fragment />}
/>
`;
